################################################################################
SHORT DESCRIPTION:
################################################################################
Fix string handling issues in blktap2 identified by Zeus's C compiler

################################################################################
LONG DESCRIPTION:
################################################################################

################################################################################
CHANGELOG
################################################################################
Author: Christopher Clark, <christopher.w.clark@gmail.com>

################################################################################
REMOVAL
################################################################################
Keep.

################################################################################
UPSTREAM PLAN
################################################################################
No, blktap2 has been removed from upstream.

################################################################################
INTERNAL DEPENDENCIES
################################################################################

################################################################################
PATCHES
################################################################################
diff --git a/tools/blktap2/drivers/block-crypto.c b/tools/blktap2/drivers/block-crypto.c
index 1a6d5eaa42..58a8135e33 100644
--- a/tools/blktap2/drivers/block-crypto.c
+++ b/tools/blktap2/drivers/block-crypto.c
@@ -175,14 +175,15 @@ find_keyfile(char **keyfile, const char *dirs,
 	*keyfile  = NULL;
 
 	while (dirs && strlen(dirs) > 0) {
-		char keydir[256] = { 0 }, path[256] = { 0 };
+		char keydir[256] = { 0 }, path[300] = { 0 };
 		struct stat st;
 		int err;
 
 		sep = strchr(dirs, ',');
 		/* get directory element */
 		if (sep == NULL) {
-			strncpy(keydir, dirs, sizeof(keydir));
+			strncpy(keydir, dirs, sizeof(keydir)-1);
+			keydir[sizeof(keydir)-1] = '\0';
 			dirs = NULL;
 		} else {
 			size_t len = sep - dirs;
@@ -191,9 +192,10 @@ find_keyfile(char **keyfile, const char *dirs,
 		}
 
 		/* check if keyfile is inside dir */
-		snprintf(path, sizeof(path),
+		snprintf(path, sizeof(path)-1,
 			 "%s/%s,aes-xts-plain,%d.key",
 			 keydir, basename, keysize);
+		path[sizeof(path)-1] = '\0';
 		err = stat(path, &st);
 		if (err == 0) {
 			/* found */
